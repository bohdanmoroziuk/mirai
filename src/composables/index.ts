export * from 'src/composables/dialog';
