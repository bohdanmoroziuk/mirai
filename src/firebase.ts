import { initializeApp } from 'firebase/app';
import { getFirestore } from 'firebase/firestore';

import config from 'src/config/firebase';

export const app = initializeApp(config);

export const firestore = getFirestore(app);
