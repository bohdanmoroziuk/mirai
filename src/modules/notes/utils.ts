import { date } from 'quasar';

export const formatDate = (value: number | string | Date) => date.formatDate(value, 'DD.MM.YYYY HH:mm');
