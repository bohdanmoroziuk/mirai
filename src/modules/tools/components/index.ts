export { default as ToolCard } from 'src/modules/tools/components/ToolCard.vue';
