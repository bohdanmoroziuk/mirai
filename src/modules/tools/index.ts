export * from 'src/modules/tools/types';
export * from 'src/modules/tools/routes';
export * from 'src/modules/tools/stores';
export * from 'src/modules/tools/services';
export * from 'src/modules/tools/components';
