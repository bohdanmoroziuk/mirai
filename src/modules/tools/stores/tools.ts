import { uid } from 'quasar';
import { defineStore } from 'pinia';

import {
  Tool,
  ToolId,
  ToolBody,
  toolsService,
} from 'src/modules/tools';

export interface ToolsStoreState {
  tools: Tool[];
}

export const createTool = (body: ToolBody): Tool => ({
  id: uid(),
  ...body,
});

export const useToolsStore = defineStore('tools', {
  persist: true,
  state: (): ToolsStoreState => ({
    tools: [],
  }),
  actions: {
    async getTools() {
      this.tools = await toolsService.getTools();
    },
    async addTool(body: ToolBody) {
      const tool = createTool(body);

      await toolsService.addTool(tool);

      this.tools = [tool, ...this.tools];
    },
    async deleteTool(id: ToolId) {
      await toolsService.deleteTool(id);

      this.tools = this.tools.filter((tool) => tool.id !== id);
    },
  },
});
