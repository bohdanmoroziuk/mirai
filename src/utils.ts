export { getLinkPreview } from 'link-preview-js';

export const validateUrl = (url: string) => {
  const regex = /(http(s)?:\/\/.)?(www\.)?[-a-zA-Z0-9@:%._+~#=]{2,256}\.[a-z]{2,6}\b([-a-zA-Z0-9@:%_+.~#?&//=]*)/gi;

  return regex.test(url);
};

export const openExternalUrl = (url: string) => {
  window.open(url, '_blank');
};
