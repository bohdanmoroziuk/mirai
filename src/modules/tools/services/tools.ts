import {
  doc,
  setDoc,
  getDocs,
  deleteDoc,
  collection,
} from 'firebase/firestore';

import { firestore } from 'src/firebase';

import { Tool, ToolId } from 'src/modules/tools';

export const TOOLS_COLLECTION_NAME = 'tools';

export const toolsService = {
  async getTools() {
    const snapshot = await getDocs(collection(firestore, TOOLS_COLLECTION_NAME));

    const tools = snapshot.docs.map((doc) => doc.data() as Tool);

    return tools;
  },
  async addTool(tool: Tool) {
    await setDoc(doc(firestore, TOOLS_COLLECTION_NAME, tool.id), tool);
  },
  async deleteTool(id: ToolId) {
    await deleteDoc(doc(firestore, TOOLS_COLLECTION_NAME, id));
  },
};
