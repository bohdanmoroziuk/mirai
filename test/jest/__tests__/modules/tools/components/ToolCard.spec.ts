import { installQuasarPlugin } from '@quasar/quasar-app-extension-testing-unit-jest';
import {
  beforeEach,
  describe,
  expect,
  jest,
  it,
} from '@jest/globals';
import { mount } from '@vue/test-utils';

import { openExternalUrl } from 'src/utils';
import { ToolCard } from 'src/modules/tools';

installQuasarPlugin();

const mockTool = {
  id: '43rfdsfdsfdfsdf',
  name: 'React Official Website',
  url: 'https://uk.reactjs.org/',
};

jest.mock('src/utils', () => ({
  openExternalUrl: jest.fn(),
}));

beforeEach(() => {
  jest.clearAllMocks();
});

describe('ToolCard', () => {
  it('renders successfully', () => {
    const wrapper = mount(ToolCard, {
      props: {
        tool: mockTool,
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders correctly', () => {
    const wrapper = mount(ToolCard, {
      props: {
        tool: mockTool,
      },
    });

    expect(wrapper.text()).toContain(mockTool.name);
  });

  it('opens the url on click', async () => {
    const wrapper = mount(ToolCard, {
      props: {
        tool: mockTool,
      },
    });

    await wrapper.trigger('click');

    expect(openExternalUrl).toHaveBeenCalledWith(mockTool.url);
  });
});
