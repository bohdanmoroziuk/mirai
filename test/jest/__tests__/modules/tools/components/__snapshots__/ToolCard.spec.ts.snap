// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToolCard renders successfully 1`] = `
<div class="q-card q-card--bordered q-card--flat no-shadow cursor-pointer">
  <div class="q-card__section q-card__section--vert q-px-xs q-py-md">
    <div class="flex justify-center"><img class="image" src="https://cdn.quasar.dev/logo-v2/svg/logo.svg" alt="React Official Website"></div>
  </div>
  <div class="q-card__section q-card__section--vert q-pa-xs q-pl-md">
    <div class="flex items-center no-wrap">
      <div class="column"><span class="text-primary">React Official Website</span>
        <!--v-if-->
      </div>
      <div class="q-space"></div><button class="q-btn q-btn-item non-selectable no-outline q-btn--flat q-btn--round text-red q-btn--actionable q-focusable q-hoverable" tabindex="0" type="button"><span class="q-focus-helper"></span><span class="q-btn__content text-center col items-center q-anchor--skip justify-center row"><i class="q-icon notranslate material-icons" aria-hidden="true" role="img">delete</i></span></button>
    </div>
  </div>
</div>
`;
